module4-authentication-and-security/r1-securing-backend-app/README.md

```diff
@@ -0,0 +1,141 @@
+
+```js
+const crypto = require("crypto");
+const size = 20; // bytes (or 40 hex digits)
+
+const token = crypto.randomBytes(size).toString("hex");
+console.log(token); // example output: 5d4225087676cf60670144e37b1694355a2e24d2
+```
+
+You can change the column `user_email` to `token` and every time a user comes in, you can ask them for the token, or allow them to generate a token. If they are new users, they can generate a new token, and save it, then start creating their lists. When they visit again, they enter that token to get their lists back.
+
+![Bingo!](https://media.giphy.com/media/W5dBnWmcGDABi/giphy.gif)
+
+Bingo! You've just created your first, very adequate authentication layer! You've secured your app and **slimmed the chances** of brute-forcing user data! Many apps today use this pattern to secure data. However, keep in mind that you usually need other features to track, like user email, name, age, etc. So such a simplistic approach wouldn't suffice. That's when this token is moved into another table and tracked with a `user_id` instead.
+
+## Why do we need passwords
+
+The token system developed above is very robust, but it isn't always an acceptable security measure (at least not alone). That's because, if a user loses the token, it is next to impossible (especially if it was generated with CSPRNGs with reasonable size) to gain access back to their lists using today's technologies.
+
+So, maybe instead of a long random token, use a password? We can ask the user when they land on our page for a passcode (aka password). And we check the database if we can find a list where `password = USER_SUBMITTED_PASSCODE` then display them. If not, that means they are a new user.
+
+Indeed this works, passwords are okay and a perfectly viable and widespread approach. Keeping in mind that users tend to use easy to remember words, dates, phone numbers, and the like. This makes passwords vulnerable to brute-force attacks. It would be relatively harder than brute-forcing emails, but still easy to guess, and very easy for computers to break by checking every possible combination of words, dates, numbers, and even relatively harder passwords using a technology called **rainbow tables**.
+
+Hence, there are two things to remember about passwords:
+They require validation to prevent brute force attacks (thus preventing easy to guess passwords)
+Use salt and hash to prevent rainbow attacks -- point to a code example here or include your own, and then this needs an in-depth explanation because this is probably the most important section that new people will forget about writing passwords
+
+## Security Checklist
+
+### Storing Passwords
+
+**Never** save the password as plain text in a database. Always use industry-standard cryptographic algorithms like bcrypt to hash and salt passwords. More on that later.
```

Comment:
```suggestion
**Never** save the password as plain text in a database. Always use industry-standard cryptographic algorithms like bcrypt to hash and salt passwords. This will be discussed further in a later section.
```